### PR TITLE
feat!: Record retirement location for ecocredit module (#328)

### PR DESCRIPTION
**Pull request by ruhatch**
## Description

Closes: #328 

This PR introduces retirement locations for ecocredits. The retirement location is the location of the beneficiary or buyer of the retired credits. It is a string of the form `<country-code>[-<sub-national-code>[-<postal-code>]]`, with the first two fields conforming to the ISO 3166-2 standard. The location is emitted during retirement events and can therefore be tracked by an indexer.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/v1alpha1/events.proto [hunk 1]
```diff
@@ -66,4 +66,10 @@ message EventRetire {
 
   // units is the decimal number of credits that have been retired.
   string units = 3;
+
+  // location is the location of the beneficiary or buyer of the retiring
+  // credits. It is a string of the form
+  // <country-code>[-<sub-national-code>[-<postal-code>]], with the first two fields
+  // conforming to ISO 3166-2
```
Comment (amaurym) [inline proto/regen/ecocredit/v1alpha1/events.proto:73]:
```suggestion
  // conforming to ISO 3166-2.
```

File: x/ecocredit/server/msg_server.go [hunk 2]
```diff
@@ -2,12 +2,14 @@ package server
...
 import (
 	"fmt"
+	"strings"
 
 	"github.com/regen-network/regen-ledger/types"
 
 	"github.com/cockroachdb/apd/v2"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/TheBookPeople/iso3166"
```
Comment (amaurym) [inline x/ecocredit/server/msg_server.go:12]: this is GPLv3, I don't think we can use it here unfortunately

File: x/ecocredit/server/testsuite/suite.go [hunk 3]
```diff
@@ -157,25 +175,47 @@ func (s *IntegrationTestSuite) TestScenario() {
...
-			name:      "can't use more than 7 decimal places",
-			toRetire:  "10.00000001",
-			expectErr: true,
+			name:               "can't use more than 7 decimal places",
+			toRetire:           "10.00000001",
+			retirementLocation: "AF",
+			expectErr:          true,
+		},
+		{
+			name:               "can't retire to an invalid country",
```
Comment (amaurym) [inline x/ecocredit/server/testsuite/suite.go:198]: we could also add one test case with empty retirementLocation

**Review by amaurym — COMMENTED** [review]
Nice work, lgtm overall. I'm just not 100% certain we can use the github.com/TheBookPeople/iso3166 library

File: x/ecocredit/server/msg_server.go (hunk 2 above)
Comment (ruhatch) [reply x/ecocredit/server/msg_server.go:12]: Oh...that's annoying! Good spot!

File: proto/regen/ecocredit/v1alpha1/events.proto [hunk 4]
```diff
@@ -66,4 +66,10 @@ message EventRetire {
 
   // units is the decimal number of credits that have been retired.
   string units = 3;
+
+  // location is the location of the beneficiary or buyer of the retiring
```
Comment (blushi) [inline proto/regen/ecocredit/v1alpha1/events.proto:70]:
```suggestion
  // location is the location of the beneficiary or buyer of the retired
```

File: x/ecocredit/server/msg_server.go [hunk 5]
```diff
@@ -345,5 +367,50 @@ func retire(ctx types.Context, store sdk.KVStore, recipient string, batchDenom b
...
+	strings := strings.Split(location, "-")
+	switch numStrings := len(strings); numStrings {
+	case 1:
+		return strings[0], "", "", nil
+	case 2:
+		return strings[0], strings[1], "", nil
+	case 3:
+		return strings[0], strings[1], strings[2], nil
+	default:
+		return "", "", "", fmt.Errorf("Location should have format <country-code>[-<region-code>[-<postal-code>]]")
```
Comment (blushi) [inline x/ecocredit/server/msg_server.go:403]: maybe we should wrap the error (for instance with `sdkerrors.ErrInvalidRequest`?)

File: x/ecocredit/server/msg_server.go [hunk 6]
```diff
@@ -345,5 +367,50 @@ func retire(ctx types.Context, store sdk.KVStore, recipient string, batchDenom b
...
+	}
+}
+
+// validatePostalCode currently checks that the postal code is not longer than
+// 10 characters, as currently the longest postal codes in the world are 10
+// characters
+func validatePostalCode(postalCode string) error {
+	if len(postalCode) > 10 {
+		return fmt.Errorf("The postal code must not be longer than 10 characters")
```
Comment (blushi) [inline x/ecocredit/server/msg_server.go:412]: ditto

File: x/ecocredit/server/msg_server.go [hunk 7]
```diff
@@ -345,5 +367,50 @@ func retire(ctx types.Context, store sdk.KVStore, recipient string, batchDenom b
 		Retirer:    recipient,
 		BatchDenom: string(batchDenom),
 		Units:      math.DecimalString(retired),
+		Location:   location,
 	})
 }
+
+// validateLocation checks that the country and region conform to ISO 3166 and
```
Comment (blushi) [inline x/ecocredit/server/msg_server.go:374]: what about extracting all the location validation related functions into a separate file? and then have a dedicated test file for it, eg location_test.go instead of msg_server_test.go

File: x/ecocredit/server/msg_server.go (hunk 2 above)
Comment (blushi) [reply x/ecocredit/server/msg_server.go:12]: very unfortunate indeed, there's also https://github.com/pariz/gountries (MIT) but doesn't seem like it's fully complete yet...
thoughts @aaronc? do you believe we should be more simplistic and just use regexp instead (see https://www.iso.org/iso-3166-country-codes.html  "WHAT IS INCLUDED IN ISO 3166?" section) without actually checking if a code really exists?

**Review by blushi — COMMENTED** [review]
Looks good overall, see small comments and follow-up on the ISO 3166 validation.

File: x/ecocredit/server/msg_server.go (hunk 2 above)
Comment (ruhatch) [reply x/ecocredit/server/msg_server.go:12]: It looks like `gountries` does have enough to make it worthwhile - it's got all the data and we can just construct our own search on the data structure it hands us.

File: x/ecocredit/server/location.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+		return err
+	}
+
+	iso3166query := gountries.New()
+
+	country, err := iso3166query.FindCountryByAlpha(countryCode)
+	if err != nil {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, err.Error())
+	}
+
```
Comment (ruhatch) [inline x/ecocredit/server/location.go:24]:
```suggestion
```

File: proto/regen/ecocredit/v1alpha1/tx.proto [hunk 9]
```diff
@@ -136,11 +144,19 @@ message MsgSendRequest {
...
-    // retired_units are the units of credits in this issuance that are
-    // effectively retired by the issuer on receipt. Decimal values are
+    // retired_units are the units of credits in this transfer that are
+    // effectively retired by the recipient on receipt. Decimal values are
     // acceptable within the precision returned by Query/Precision.
     string retired_units = 3
         [ (gogoproto.moretags) = "yaml:\"retired_units\"" ];
+
+    // retirement_location is the location of the beneficiary or buyer of the
+    // retired credits. This must be provided if retired_units is positive. It
```
Comment (aaronc) [inline proto/regen/ecocredit/v1alpha1/tx.proto:154]: Is it actually required? @blushi or @clevinson can you check with Ron?

File: x/ecocredit/server/location_test.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+			expectErr: false,
+		},
+		{
+			name:      "can't validate invalid region code",
+			location:  "BF-ZZ",
+			expectErr: true,
+		},
+		{
+			name:      "can validate valid postal code",
+			location:  "BF-BAL-1",
```
Comment (aaronc) [inline x/ecocredit/server/location_test.go:52]: I would say we want a space instead of a `-` before the postal code. What do you think?

File: x/ecocredit/server/location.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+// the postal code is valid
+func validateLocation(location string) error {
+	countryCode, regionCode, postalCode, err := parseLocation(location)
+	if err != nil {
+		return err
+	}
+
+	iso3166query := gountries.New()
+
+	country, err := iso3166query.FindCountryByAlpha(countryCode)
```
Comment (aaronc) [inline x/ecocredit/server/location.go:20]: So I'd rather not check the list of countries and sub regions. This can change with little notice and we don't want to require a hard fork to update it. So for now I'd say we should just use a regex to validate 2 alphabetic characters for country and 1-3 alphanumeric characters for subregion. That's more future proof and doesn't introduce a possibly buggy third party dependency

File: x/ecocredit/server/location.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+	default:
+		return "", "", "", sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Location should have format <country-code>[-<region-code>[-<postal-code>]]")
+	}
+}
+
+// validatePostalCode currently checks that the postal code is not longer than
+// 10 characters, as currently the longest postal codes in the world are 10
+// characters
+func validatePostalCode(postalCode string) error {
+	if len(postalCode) > 10 {
```
Comment (aaronc) [inline x/ecocredit/server/location.go:57]: This seems a little too stringent. Maybe we just validate that it's alphanumeric plus maybe dashes and not longer than 16 or 20 characters maybe?

File: x/ecocredit/server/msg_server.go (hunk 2 above)
Comment (aaronc) [reply x/ecocredit/server/msg_server.go]: Just regex please as I commented elsewhere

Comment (aaronc) [reply x/ecocredit/server/msg_server.go]: In the future we can have a dataset managed with x/params maybe but we do not want to force a hard fork

File: x/ecocredit/server/location_test.go (hunk 10 above)
Comment (ruhatch) [reply x/ecocredit/server/location_test.go:52]: Not sure! Why would you prefer one to the other?

File: x/ecocredit/server/location.go (hunk 11 above)
Comment (ruhatch) [reply x/ecocredit/server/location.go:20]: Yeah, makes sense

File: x/ecocredit/server/location.go (hunk 12 above)
Comment (ruhatch) [reply x/ecocredit/server/location.go:57]: Sure. I've just done an audit of global postcode formats and we'll want alphanumeric, plus dash, parentheses, and spaces. Some postal codes include the name of a city, so actually we'd probably want more like 30 characters. Here's a long example from Gabon: `01 LIBREVILLE–GROS-BOUQUET 03`.

File: x/ecocredit/server/location_test.go (hunk 10 above)
Comment (aaronc) [reply x/ecocredit/server/location_test.go:52]: Hmm just seems more visually clear

File: x/ecocredit/server/location.go (hunk 12 above)
Comment (aaronc) [reply x/ecocredit/server/location.go:57]: Why not be generous and say 32 or 64?

File: x/ecocredit/server/location_test.go (hunk 10 above)
Comment (ruhatch) [reply x/ecocredit/server/location_test.go:52]: Sure, I can just work the whole thing into a single regex then

**Comment by ruhatch** [comment]
Also need to add `client/tx` for locations

**Review by aaronc — COMMENTED** [review]
Let's also add the location validation logic to the `Msg.ValidateBasic` methods

File: proto/regen/ecocredit/v1alpha1/tx.proto (hunk 9 above)
Comment (blushi) [reply proto/regen/ecocredit/v1alpha1/tx.proto:154]: why wouldn't it be required in this case? As long as credits get retired, even if it's on issuance, the location should be recorded, even though this use case might not happen that often.

File: x/ecocredit/location_test.go [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+			expectErr: true,
+		},
+		{
+			name:      "can validate valid region code",
+			location:  "AF-BDS",
+			expectErr: false,
+		},
+		{
+			name:      "can validate valid region code 2",
+			location:  "BF-BAL",
```
Comment (aaronc) [inline x/ecocredit/location_test.go:42]: Let's check numeric region codes too

File: x/ecocredit/location_test.go [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+func TestValidateLocation(t *testing.T) {
+	specs := []struct {
+		name      string
+		location  string
+		expectErr bool
+	}{
+		{
+			name:      "can validate valid country code",
```
Comment (aaronc) [inline x/ecocredit/location_test.go:16]:
```suggestion
			name:      "valid country code",
```
we can drop the "can/can't validate" for better readability

File: x/ecocredit/location_test.go [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+			location:  "AF",
+			expectErr: false,
+		},
+		{
+			name:      "can validate valid country code 2",
+			location:  "BF",
+			expectErr: false,
+		},
+		{
+			name:      "can't validate invalid country code",
```
Comment (aaronc) [inline x/ecocredit/location_test.go:26]:
```suggestion
			name:      "invalid country code",
```

File: proto/regen/ecocredit/v1alpha1/tx.proto [hunk 16]
```diff
@@ -136,11 +144,19 @@ message MsgSendRequest {
...
+    // is a string of the form
+    // <country-code>[-<sub-national-code>[-<postal-code>]], with the first two
```
Comment (aaronc) [inline proto/regen/ecocredit/v1alpha1/tx.proto:156]: This needs to be updated to describe the space character and postal code format (64 char alphanumeric)

File: x/ecocredit/msgs.go [hunk 17]
```diff
@@ -60,10 +67,17 @@ func (m *MsgSendRequest) ValidateBasic() error {
...
 		}
 
-		_, err = math.ParseNonNegativeDecimal(iss.RetiredUnits)
+		retiredUnits, err := math.ParseNonNegativeDecimal(iss.RetiredUnits)
 		if err != nil {
 			return err
 		}
+
+		if !retiredUnits.IsZero() {
+			err = validateLocation(iss.RetirementLocation)
```
Comment (aaronc) [inline x/ecocredit/msgs.go:76]: @blushi @clevinson do we always require a valid location? Could someone retire credits without providing a location theoretically?

**Review by aaronc — APPROVED** [review]
A few small comments, but pre-approving. Thanks @ruhatch !

File: x/ecocredit/msgs.go (hunk 17 above)
Comment (blushi) [reply x/ecocredit/msgs.go:76]: I don't think so, this should be required as long as credits get retired

Comment (ruhatch) [reply x/ecocredit/msgs.go:76]: @blushi is it worth a final check with Ron?

**Review by amaurym — APPROVED** [review]
nice! lgtm
